---
title: 취소 토큰
date: 2024-07-27 08:00:00 +09:00
categories: [비동기 프로그래밍]
tags:
  [
    CancellationTokenSource,
    CancellationToken
  ]
---

### 개요
- 다음 메서드는 .NET에서 제공하는 비동기 메서드이다.
    ```cs
    await _httpClient.GetAsync(uri);
    ```
- HttpClient.GetAsync() 메서드는 여러 인자를 받을 수 있도록 오버로드 되어있는데 그중 하나가 CancellationToken(이하 CancelToken)이다.
    ```cs
    var cancelToken = new CancellationToken();
    await _httpClient.GetAsync(uri, cancelToken);
    ```
- 실행 시간이 50ms를 넘을 수 있는 메서드는 비동기메서드로 제공되는데, 때문에 대부분의 .NET 비동기 메서드에서 CancelToken 인자를 입력 받을 수 있도록 오버로드된 메서드를 제공한다. 즉, 실행이 완료되기까지 오랜 시간이 걸릴 가능성이 있는 메서드는 실행 중 취소할 수 있도록 설계한 것이다.

```
취소는 언제 필요할까?

- 취소는 단지 취소에만 필요한게 아니다! 부분 작업 실패에 따른 전체 작업 중단, Task.WhenAny()와 같이 두 가지 작업 중 하나의 결과만 받아도 되는 경우, 나머지 작업 취소(리소스 정리) 등에도 필요하다. 

- [예시] 하나의 파일을 5개로 쪼개 나누어 다운로드 받아야 한다고 생각하자. 전체 작업(파일 다운로드)는 5개의 작업이 모두 성공해야 성공한다. 만약 이중에서 2번째 작업이 실패했다면 나머지 1, 3, 4, 5번 작업을 계속할 필요가 없다.
```

### CancelToken 이란?
- 특정 메서드를 실행중 취소할 때 사용하는 토큰이다. 네트워크 상황에 따라 HttpClient.GetAsync() 메서드가 완료되는 시간은 달라진다. Windows 운영체제에서 GetAsync 호출 중 네트워크가 끊긴 경우, 별도의 타임아웃을 설정해 놓지 않으면 무한정 대기하는 경우도 발생한다. (만약 해당 메서드를 포함한 기능을 제공 받는 쪽이라면, 어떠한 알림도 받지 못한채 네트워크가 연결되기만을 기다려야할 것이다.)

- 이때 메서드가 CancellationToken을 전달받는다면 취소가 가능하다.
    ```cs
    var cts = new CancellationTokenSource();
    // 5초 뒤에 취소 토큰에 취소를 보낸다.
    cts.CancelAfter(TimeSpan.FromSeconds(5));

    try
    {
        // 해당 메서드가 5초내에 완료되지 않으면 OperationCanceledException 예외가 발생한다. 
        await client.GetAsync(uri, cts.Token);

        // 성공
    }
    catch (OperationCanceledException)
    {
        // 취소
        MessageBox.Show("네트워크가 불안정합니다.");
    }
    catch (Exception )
    {
        //실패
    }
    ```

### CancellationTokenSource    
- CancellationTokenSource(이하 cts)가 등장했다. CancelToken는 cts를 통해 생성되고, 직접적으로 취소를 호출하는 주체도 cts가 담당한다. 
- 이름에서 드러나듯이 Cts는 CancelToken의 소스이고, 하나의 cts에서 여러 개의 CancelToken을 생성할 수 있다. 취소를 호출되면 연결된 CancelToken를 취소 상태로 만든다.

### 왜 취소 기능을 분리했을까?
- Cts 는 취소 요청을 보내는 전송자, CancelToken은 취소 요청을 수신하는 수신자 역할로 분리하여 취소 기능을 모듈화한다.
- cts를 갖는 쪽에서만 취소 권한을 갖고, cancelToken을 받은 쪽은 취소 요청만 받도록 분리한 것이다. (취소 토큰을 전달받은 곳에서 제 맘대로 취소를 해버리길 바라진 않을 것이다.)

### CancellationTokenSource.CreateLinkedTokenSource()
- 취소 요청이 발생할 수 있는 작업끼리 연결할 때 사용.
- a 기능과 b 기능이 독립적으로 취소 기능을 구현할 때, 논리에 따라 a 작업이 b작업을 포함해야 하는 경우, a작업의 취소는 b작업의 취소로 이루어져야 한다. 이때 b의 Cts를 a로부터 생성하면 a 의 취소가 b에도 신호를 보낼 수 있게 된다.

- 단, 이때 b의 cts 취소 요청을 보내도 a는 취소되지 않는다!
```cs

// cancelToken으로부터 새로운 CancellationTokenSource 생성
var anotherCts = CancellationTokenSource.CreateLinkedTokenSource(cancelToken);

(예시 생각중...)
public async Task DoSomething()
{

}

public async Task RequestSomethingAsync()
{

}
```

### (주의) 취소 메서드를 별도로 구성하지마라.
- 가끔 메서드의 취소를 별도의 메서드로 구성하여 제공하는 경우가 있다. 다음은 다운로드 기능을 API를 제공하는 예시 코드다.
    ```cs
    // SimpleDownloader.cs

    public async Task DownloadAsync()
    {
        ...

        // 다운로드
        var respMsg = await _client.GetAsync(uri, _cts.Token);

        ...
    }

    public async Task CancelDownloadAsync()
    {
        // 다운로드 취소
        _cts?.Cancel(true);
    }
    ```

- 언뜻보면 문제 없는 것 같지만, 사용자 입장에서 보면 문제가 드러난다. 
1. 다운로드 완료 결과를 별도로 조회해야 한다.
```cs
var result = await _simpleDownloader.DownloadAsync();
    
// 다운로드가 취소 없이 완료되었는지 알기 위해서는 별도의 결과를 조회해야 한다.
if (result.IsSuccessful)
{
    // 성공 처리
}
else if (result.IsCanceled)
{
    // 취소 처리
}
```
- 주관적인 의견이지만, 나는 로직에서 성공 실패를 조회하는 것을 최대한 지양한다. 모든 로직은 성공을 가정하여 간결하게 작성하고, 예외는 catch문으로 분리해야 한다고 생각하기 때문이다. (DownloadAsync() 메서드가 완료되었으면 성공한 거지, 도대체 왜 추가로 결과를 조회하게 만든단 말인가.)

- 내 의견에는 동조하지만, 이정도는 감수할 수 있다고 말하는 개발자는 많을 것이다. 그러나 훨씬 심각한 문제가 남아있다. 

2. 다운로드 취소 처리가 중복될 수 있다.
- 앞에서는 다운로드 취소 요청 이후 로직을 다운로드 메서드 이후에 작성했다. 그러나 사용자는 다운로드 취소 메서드 이후에 로직을 작성할 수도 있다.
```cs
await CancelDownloadAsync();

// 여기서 다운로드 취소 처리
```
- 사용자는 다운로드 취소 처리 로직을 DownloadAsync() 메서드 이후에 작성해야 하는지, CancelDownloadAsync()에 작성해야 하는지 혼동할 것이고, 중복으로 작성해야 한다고 느낄 수도 있다.

> 1, 2와 같은 문제가 발생하는 핵심은 다운로드 메서드 호출 결과(성공/실패/취소)를 한 곳에서 처리하지 못했기 때문이다. 
> 다운로드 결과는 성공, 실패(취소) 중 하나이기 때문에, 요청도 한곳에서 이루어져야 처리가 편하다.

#### 정리
- 취소 기능을 구현할 때는 별도의 API 메서드로 구성하지 말고, 메서드가 취소 토큰을 전달받을 수 있도록 오버로드하여 제공해라.







